src/WebJobs.Script/Diagnostics/FileTraceWriter.cs

```diff
@@ -216,29 +210,91 @@ protected virtual void AppendLine(string line)
             _logBuffer.Enqueue(line);
         }
 
-        private string FormatMessage(string message)
-            => string.Format(CultureInfo.InvariantCulture, "{0} {1}", DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fff", CultureInfo.InvariantCulture), message.Trim());
+        private string FormatMessage(string message) => $"{DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fff")} {message.Trim()}";
 
         private void OnFlushLogs(object sender, ElapsedEventArgs e)
         {
             Flush();
         }
 
-        internal void SetNewLogFile()
+        /// <summary>
+        /// Called to set/create the current log file that will be used. Also handles purging of old
+        /// log files.
+        /// </summary>
+        internal void SetLogFile()
+        {
+            if (!_logDirectory.Exists)
+            {
+                _logDirectory.Create();
+            }
+
+            // Check to see if there is already a log file we can use
+            bool newLogFileCreated = false;
+            var logFile = GetCurrentLogFile();
+            if (logFile != null)
+            {
+                _currentLogFileInfo = logFile;
+            }
+            else
+            {
+                lock (_syncLock)
+                {
+                    // we perform any file create operations in a lock to avoid
+                    // race conditions with multiple instances of this class on
+                    // multiple threads, where multiple log files might be created
+                    // unnecessarily
+                    logFile = GetCurrentLogFile();
+                    if (logFile != null)
+                    {
+                        _currentLogFileInfo = logFile;
+                    }
+                    else
+                    {
+                        // create a new log file
+                        // we include a machine identifier in the log file name to ensure we don't have any
+                        // log file contention between scaled out instances
+                        var timestamp = DateTime.UtcNow.ToString("yyyy-MM-ddTHH-mm-ssK");
+                        string filePath = Path.Combine(_logFilePath, $"{timestamp}-{_instanceId}.log");
+                        _currentLogFileInfo = new FileInfo(filePath);
+                        _currentLogFileInfo.Create().Close();
```

Comment:
This is a change - we now create the empty log file as part of setting it. This addresses the race condition by allowing the second instance to discover the log file created by the first once it gets the lock.